sled-agent/config-reconciler/src/dataset_serialization_task.rs

```diff
@@ -399,9 +456,140 @@ impl DatasetTask {
         Ok(props.into_iter().map(From::from).collect())
     }
 
+    async fn datasets_report_orphans<T: ZfsImpl>(
+        &mut self,
+        datasets: IdMap<DatasetConfig>,
+        currently_managed_zpools: Arc<CurrentlyManagedZpools>,
+        zfs: &T,
+    ) -> anyhow::Result<IdOrdMap<OrphanedDataset>> {
+        let mut orphaned_datasets = IdOrdMap::new();
+
+        let datasets_of_interest = currently_managed_zpools
+            .iter()
+            .flat_map(|zpool| {
+                [
+                    // We care about the direct children of the zpool itself.
+                    // (This will include any unencrypted datasets, such as
+                    // `crucible`.)
+                    zpool.to_string(),
+                    // We also care about the direct children of the encrypted
+                    // dataset. (This includes non-crucible Omicron zone durable
+                    // datasets.)
+                    format!("{zpool}/{CRYPT_DATASET}"),
+                    // We _don't_ need to check children of
+                    // `{zpool}/{ZONE_DATASET}`: these are all transient, and
+                    // should be destroyed/created on demand.
```

Comment:
To be super explicit, what's deleting these datasets today? Is it the deletion of the zone?